### PR TITLE
make optional fields optional when deserializing

### LINKED ISSUE #229: optional fields are required when deserializing
I got an error:
```
Error: DeserializeError { source: Error("missing field `resourceAccessRules`", line: 1, column: 504),
```

when running the storage example:
```
services> cargo run --example storage_account_list
```

Instead of 
``` rust
#[serde(rename = "resourceAccessRules", skip_serializing_if = "Vec::is_empty")]
```
it should be
``` rust
#[serde(rename = "resourceAccessRules", default, skip_serializing_if = "Vec::is_empty")]
```

https://serde.rs/field-attrs.html

This is a very simple bug fix once #206 gets merged.

### PR DESCRIPTION
**Pull request by ctaggart**
Fix #229.

### CODE REVIEW

File: services/autorust/codegen/src/codegen.rs [hunk 1]
```diff
@@ -254,9 +254,9 @@ impl CodeGen {
             } else {
                 if !is_required {
                     if is_vec {
-                        serde_attrs.push(quote! { skip_serializing_if = "Vec::is_empty"});
+                        serde_attrs.push(quote! { default, skip_serializing_if = "Vec::is_empty"});
```
Comment (heaths) [inline services/autorust/codegen/src/codegen.rs:257]: 1. What happens with `Option<Vec<T>>`?
2. With some models, an empty JSON array is valid and should be treated differently than null. In fact, generally, explicit nulls are fairly rare (except with services that use the Microsoft.OData library on the backend). Admittedly, this is a problem we haven't quite figured out with the other SDKs yet, and recommend people describe their swagger using `x-nullable: true` if a value could be an explicit null. Are you taking that into account already? See https://github.com/Azure/autorest/tree/master/docs/extensions#x-nullable for details.

Comment (ctaggart) [reply services/autorust/codegen/src/codegen.rs:257]: - I added https://github.com/Azure/azure-sdk-for-rust/issues/234 to support `x-nullable`. It is not currently used.
- In this case, the field is not required `!is_required`. A `Option<Vec<T>>` is not needed. A `Vec<T>` is used.
-  If the field is required, the vector will be serialized, even if empty.

Comment (ctaggart) [reply services/autorust/codegen/src/codegen.rs:257]: Here, `default` only affects the deserialization. The optional field will be set to an empty Vec if not in the json.

Comment (heaths) [reply services/autorust/codegen/src/codegen.rs:257]: @JeffreyRichter for the control plane, this seems sufficient. Do you agree? Off hand, I'm not even sure if the control plane uses x-nullable. I only know that some of the data plane do.

Comment (JeffreyRichter) [reply services/autorust/codegen/src/codegen.rs:257]: I can tell you what we expect for new services that follow our guidelines:

- A service should never return a null back to a client; either the field is there it is isn't there.
- A client should only ever send a null inside a PATCH's JSON merge patch payload if the client wants the service to delete the field. If a field's value is null, then the field is NOT sent over the wire. To send a null for JSON merge patch, the customer must explicitly set the field to some special sentinel value that tells the serializer to send the field with a value of null - we can worry about how to best do this in Rust later (it is a corner case).
- A client should never send a null in a payload in any other situation.

Comment (ctaggart) [reply services/autorust/codegen/src/codegen.rs:257]: Thanks for the guidelines! This pull request matches them. Yes, let's worry about sending a `null` value in a payload later. Can someone please approve this pull request? It fixes a lot of the deserialization. I've tested it with storage.

**Review by heaths — APPROVED** [review]
Based on services/autorust/codegen/src/codegen.rs and spot checks of a few of the generated files (all the rest besides the aforementioned file appeared to be generated, but please correct me if I'm wrong), LGTM.

File: services/autorust/codegen/src/codegen.rs (hunk 1 above)
Comment (heaths) [reply services/autorust/codegen/src/codegen.rs:257]: Thanks, @JeffreyRichter. A lot of clients using the Microsoft OData library do send back explicit `null` in a lot of cases - often not attributing the property with `x-nullable: true`, so I wonder if working with the owners to change that behavior (or at least make it opt-out) could go a long way to improving that situation. For example, I see a lot of explicit `null` in Cognitive services.

Obviously this shouldn't hold up this PR, so I've approved.

**Comment by ctaggart** [comment]
That is correct. `codegen.rs` only adds `default` on two lines. Everything else is generated. Thanks for the approval.

**Pull request merged**